### PR TITLE
#1559 #1560 #1561 Document Modal part 1

### LINKED ISSUE #1559: Document modal -- implement for FileUpload
Part of Epic #1549 

In the first instance, we just want to get it going on the FileUpload element. Once we're happy with that it'll be easier to add elsewhere.

### PR DESCRIPTION
**Pull request by CarlosNZ**
Fix #1559 & #1560 & #1561

Creates "Document Modal" and implements it in the following places:
- **Application view** (in forms)
- **Summary View** (for reviewer/summary page)
- **Database** views
- "**Notes**" tab (review page)
- "**Documents**" tab (review page)
- **Previews** (I know we said we would probably not do this one, but it actually looks pretty good and it was easy, so.._)

Still to do (might update PR, might do in new issues if substantial):

- [x] Downloaded file has wrong name
-  ~~Simple "Zoom" functionality~~ (don't think we need this)
- [x] Cache file so it doesn't take time to re-download a file we just uploaded

Note: you'll need to turn on the preference `showDocumentModal: true` in system prefs (web section) to get the modal functionality.

**Viewing a PDF:**

Screenshot 2023-10-27 at 2 35 38 PM

**Viewing an image file:**

Screenshot 2023-10-27 at 2 35 52 PM

**And when you try to view a file that is not supported (i.e. not pdf or image):**

Screenshot 2023-10-27 at 2 36 08 PM

See inline comments for further detail

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "react-dom": "^18.2.0",
     "react-ga4": "^2.1.0",
     "react-markdown": "^5.0.3",
+    "react-pdf": "5.x",
```
Comment (CarlosNZ) [inline package.json:43]: This is the latest version of this package that I could get to work with our current setup.

File: src/containers/Dev/Snapshots.tsx [hunk 2]
```diff
@@ -248,7 +230,17 @@ const Snapshots: React.FC = () => {
               className="clickable blue"
               onClick={async () => {
                 showToast({ title: 'Download started...', timeout: 2000 })
-                await downloadSnapshot(filename)
+                await downloadFile(
```
Comment (CarlosNZ) [inline src/containers/Dev/Snapshots.tsx:233]: Made a generic utility function "download file" so can be re-used in Document viewer (and anywhere else we might need it)

File: src/contexts/SystemPrefs.tsx [hunk 3]
```diff
@@ -3,33 +3,44 @@ import { LanguageOption } from './Localisation'
...
-  }
+  preferences: Preferences
   languageOptions?: LanguageOption[]
   latestSnapshot?: string
   loading: boolean
   error?: Error
 }
 
+const defaultPrefs: Preferences = {
+  showDocumentModal: false,
```
Comment (CarlosNZ) [inline src/contexts/SystemPrefs.tsx:27]: Might change this to `true`. Will see what people think.

File: src/formElementPlugins/fileUpload/localisation.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
   "ERROR_UPLOAD_PROBLEM": "Problem uploading file",
   "BUTTON_CLICK_TO_UPLOAD": "Click to upload",
   "BUTTON_UPLOAD_ANOTHER": "Upload another",
+  "FILE_UPLOADING": "Uploading...",
```
Comment (CarlosNZ) [inline src/formElementPlugins/fileUpload/localisation.json:9]: Wasn't localised, oops

File: src/formElementPlugins/fileUpload/src/ApplicationView.tsx [hunk 5]
```diff
@@ -50,6 +53,7 @@ const ApplicationView: React.FC<ApplicationViewProps> = ({
     fileExtensions,
     fileSizeLimit,
     default: defaultValue,
+    showDocumentModal = preferences.showDocumentModal,
```
Comment (CarlosNZ) [inline src/formElementPlugins/fileUpload/src/ApplicationView.tsx:56]: The system-wide preference for "showDocumentModal" can be overridden for each FileUpload element if desired.

File: src/utils/helpers/utilityFunctions.ts [hunk 6]
```diff
@@ -78,3 +78,13 @@ export const fileSizeWithUnits = (size: number): string => {
   const sizeInGB = size / 1_000_000_000
   return `${parseInt(String(sizeInGB * 100)) / 100} GB`
 }
+
+// Force a file download
+export const downloadFile = async (url: string, filename: string, fetchOptions: object = {}) => {
```
Comment (CarlosNZ) [inline src/utils/helpers/utilityFunctions.ts:83]: Abstracted from snapshots.

File: src/utils/hooks/useDebouncedCallback.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { useCallback, DependencyList } from 'react'
+
+export const useDebounceCallback = <T extends (...args: any[]) => any>(
```
Comment (CarlosNZ) [inline src/utils/hooks/useDebouncedCallback.ts:3]: This hook is copied from open mSupply.

File: src/utils/hooks/useWindowDimensions.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { useState, useEffect } from 'react'
+
+const getWindowDimensions = () => {
```
Comment (CarlosNZ) [inline src/utils/hooks/useWindowDimensions.ts:3]: This hook mostly taken from online source. The Pdf viewer requires size attributes in pixels, so we need to get the current viewport width to calculate a reasonable width for the modal.

File: tsconfig.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     /* Basic Options */
-    "target": "es5" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
+    "target": "ES2015" /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */,
```
Comment (CarlosNZ) [inline tsconfig.json:4]: Required to the get react-pdf package working. Still builds okay and runs on demo server fine.

**Comment by regotaina** [comment]
Hey @CarlosNZ ! I will do a few more tests, it looks nice on desktop and it doesn't take long to load!

On mobile it looks wobbly, I imagine we will address this at a second stage:

telegram-cloud-photo-size-1-5041837769650973642-y

Good work! :)

**Comment by CarlosNZ** [comment]
> On mobile it looks wobbly, I imagine we will address this at a second stage:

Thanks, I *think* this is actually because the underlying page is not behaving nicely on Mobile and the modal will be okay once we fix that. But will need to test it in other contacts to confirm that.

Thanks.

**Comment by CarlosNZ** [comment]
> Thanks, I _think_ this is actually because the underlying page is not behaving nicely on Mobile and the modal will be okay once we fix that. But will need to test it in other contacts to confirm that.

Yeah, confirmed. If you try it on an application page (with okay mobile formatting), the modal seems (mostly) good:
Screenshot 2023-10-30 at 5 34 48 PM

File: tsconfig.json (hunk 9 above)
Comment (guatemartin) [reply tsconfig.json:4]: That's a bit of a shame...

**Review by guatemartin — APPROVED** [review]
Yeah, lots of changes here.  Best approach will be to implement and test in use.

**Comment by CarlosNZ** [comment]
> Best approach will be to implement and test in use.

Yup, it's been in use on 50008 demo for a couple of weeks now :)

**Pull request merged**